overlord/fdestate/handlers_test.go

```diff
@@ -0,0 +1,388 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package fdestate_test
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"sort"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/overlord/fdestate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/secboot/keys"
+	"github.com/snapcore/snapd/strutil"
+)
+
+func (s *fdeMgrSuite) mockCurrentKeys(c *C, rkeys, unlockKeys []fdestate.KeyslotRef) {
+	var dataUnlockKeyNames, saveUnlockKeyNames []string
+	if len(unlockKeys) == 0 {
+		dataUnlockKeyNames = []string{"default", "default-fallback"}
+		saveUnlockKeyNames = []string{"default", "default-fallback"}
+	} else {
+		for _, ref := range unlockKeys {
+			switch ref.ContainerRole {
+			case "system-data":
+				dataUnlockKeyNames = append(dataUnlockKeyNames, ref.Name)
+			case "system-save":
+				saveUnlockKeyNames = append(saveUnlockKeyNames, ref.Name)
+			default:
+				c.Errorf("unexpected unlock key slot reference: %s", ref.String())
+			}
+		}
+	}
+
+	var dataRecoveryKeyNames, saveRecoveryKeyNames []string
+	if len(rkeys) == 0 {
+		dataRecoveryKeyNames = []string{"default-recovery"}
+		saveRecoveryKeyNames = []string{"default-recovery"}
+	} else {
+		for _, ref := range rkeys {
+			switch ref.ContainerRole {
+			case "system-data":
+				dataRecoveryKeyNames = append(dataRecoveryKeyNames, ref.Name)
+			case "system-save":
+				saveRecoveryKeyNames = append(saveRecoveryKeyNames, ref.Name)
+			default:
+				c.Errorf("unexpected recovery key slot reference: %s", ref.String())
+			}
+		}
+	}
+
+	s.mockDeviceInState(&asserts.Model{}, "run")
+
+	s.AddCleanup(fdestate.MockDisksDMCryptUUIDFromMountPoint(func(mountpoint string) (string, error) {
+		switch mountpoint {
+		case filepath.Join(dirs.GlobalRootDir, "run/mnt/data"):
+			return "data", nil
+		case dirs.SnapSaveDir:
+			return "save", nil
+		}
+		panic(fmt.Sprintf("missing mocked mount point %q", mountpoint))
+	}))
+
+	s.AddCleanup(fdestate.MockSecbootListContainerUnlockKeyNames(func(devicePath string) ([]string, error) {
+		switch devicePath {
+		case "/dev/disk/by-uuid/data":
+			return dataUnlockKeyNames, nil
+		case "/dev/disk/by-uuid/save":
+			return saveUnlockKeyNames, nil
+		default:
+			return nil, fmt.Errorf("unexpected devicePath %q", devicePath)
+		}
+	}))
+
+	s.AddCleanup(fdestate.MockSecbootListContainerRecoveryKeyNames(func(devicePath string) ([]string, error) {
+		switch devicePath {
+		case "/dev/disk/by-uuid/data":
+			return dataRecoveryKeyNames, nil
+		case "/dev/disk/by-uuid/save":
+			return saveRecoveryKeyNames, nil
+		default:
+			return nil, fmt.Errorf("unexpected devicePath %q", devicePath)
+		}
+	}))
+}
+
+func (s *fdeMgrSuite) TestDoAddRecoveryKeys(c *C) {
```

Comment:
you can check the style of tests we have in snapstate for handlers, but usually we have idempotency tests as well, about what happens if the method is called twice

Reply:
Good point, thank you! `doRenameKeys` and `doRemoveKeys` are idempotent by design, I will add corresponding tests.

However `doAddRecoveryKeys` always fails in subsequent calls due to the fact that the recovery key is consumed in the first run or lost  due to a reboot.

Reply:
I figure this would be similar to when removing snap with purge which is not reversible, I will check what we do there.